db/kvstorewithcache.go

```diff
@@ -0,0 +1,150 @@
+package db
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	"github.com/iotexproject/go-pkgs/cache"
+	"github.com/iotexproject/iotex-core/db/batch"
+)
+
+// kvStoreWithCache is an implementation of KVStore, wrapping kvstore with LRU caches of latest states
+type kvStoreWithCache struct {
+	mutex       sync.RWMutex // lock for stateCaches
+	store       KVStore
+	stateCaches map[string]*cache.ThreadSafeLruCache // map having lru cache to store current states for speed up
+	cacheSize   int
+}
+
+// NewKvStoreWithCache wraps kvstore with stateCaches
+func NewKvStoreWithCache(kvstore KVStore, cacheSize int) KVStore {
+	return &kvStoreWithCache{
+		store:       kvstore,
+		stateCaches: make(map[string]*cache.ThreadSafeLruCache),
+		cacheSize:   cacheSize,
+	}
+}
+
+// Start starts the kvStoreWithCache
+func (kvc *kvStoreWithCache) Start(ctx context.Context) error {
+	return kvc.store.Start(ctx)
+}
+
+// Stop stops the kvStoreWithCache
+func (kvc *kvStoreWithCache) Stop(ctx context.Context) error {
+	kvc.mutex.Lock()
+	defer kvc.mutex.Unlock()
+	for _, sc := range kvc.stateCaches {
+		sc.Clear()
+	}
+	return kvc.store.Stop(ctx)
+}
+
+// Put inserts a <key, value> record into stateCaches and kvstore
+func (kvc *kvStoreWithCache) Put(namespace string, key, value []byte) (err error) {
+	kvc.updateStateCachesIfExists(namespace, key, value)
+	return kvc.store.Put(namespace, key, value)
```

Comment:
I think we should Put into store first, and then the cache, in case that "kvc.store.Put" fails.

Reply:
Good point, fixed on latest commit